Update marketing button styles

tobiasahlin opened the pull request:
This PR aligns the marketing button styles to those used on GitHub.com, adds a purple variant (`.btn-purple-mktg`), a `.btn-sm-mktg` modifier for small buttons, as well as updates the `.btn-large-mktg` utility to use the more common `-lg` modifier: `.btn-lg-mktg`.

This PR is dependent on `primitives` changes in https://github.com/primer/primitives/pull/76.

Preview of variants:

<img width="960" alt="Screen Shot 2021-04-19 at 17 41 27" src="https://user-images.githubusercontent.com/211284/115264780-e52a6980-a136-11eb-8ea2-609057f803a5.png">


Preview of sizes:

<img width="952" alt="Screen Shot 2021-04-19 at 17 41 36" src="https://user-images.githubusercontent.com/211284/115264792-e9ef1d80-a136-11eb-87c4-40fbe7ebfb84.png">

## Migration guide

- Replace `.btn-large-mktg` with `.btn-lg-mktg`


/cc @primer/ds-core

trosage commented on src/marketing/buttons/button.scss:
```diff
@@ -1,117 +1,92 @@
-  }
+.btn-sm-mktg {
+  // stylelint-disable-next-line primer/spacing
+  padding: rem(10px) rem($spacer-4) rem(12px);
```
`padding: rem(10px) rem($spacer-4) rem(13px);`

I bumped up the bottom padding here slightly for when it uses Alliance so the padding is optically correct.

simurai replied:
👍  Primer CSS uses the "system" font, so might wanna have the same top/bottom padding so it doesn't look mis-aligned? And maybe also a less bold `font-weight`?

tobiasahlin replied:
I tweaked the small ones with more spacing below, and also:
- Reduced spacing on the sides, to $spacer-3, for better balance
- Added one pixel of spacing to the `lg` size modifier

Since these are primarily for use with Alliance, I think it's more important that they work with that font for now. In the cases when they are not used with Alliance (I'm not sure if they should be at all), the padding can be overridden. However, I think we should consider this a bug with how Alliance is set up, and we should investigate if we can export Alliance with different settings to prevent us from "hacking" the vertical alignment in this way.

Preview:

<img width="679" alt="Screen Shot 2021-04-22 at 11 30 16" src="https://user-images.githubusercontent.com/211284/115691767-a4149e00-a35e-11eb-9569-83930e9cd6f5.png">

simurai replied:
> However, I think we should consider this a bug with how Alliance is set up, and we should investigate if we can export Alliance with different settings to prevent us from "hacking" the vertical alignment in this way.

Yeah, I guess when we update the [Marketing Typorgraphy](https://primer.style/css/utilities/marketing-type) there is no way around to try include Alliance. Is Alliance an open source font or is a license needed to use it? [Looks like the later](https://degarism.com/Alliance). So maybe we could include it in the marketing font stack and then document that all the marketing components are optimized for Alliance. But maybe out of scope for this PR.

simurai reviewed (commented):
One question before going into more details: Can/should these marketing buttons be used on pages that support color modes? 

![buttons](https://user-images.githubusercontent.com/378023/115331904-c0cf9b00-a1d1-11eb-9773-732b760b44cf.gif)

Maybe fine except for the `btn-mktg btn-outline-mktg` that could use some more contrast in dark mode?

In case they are only meant for light mode, maybe it needs to be documented somewhere. And we need a way to not show the color modes toggle in the docs?

simurai commented:
Looked into it a bit more and how I understood it: We have 3 kinds of marketing buttons.

- [Primer CSS Marketing](https://primer.style/css/components/marketing-buttons) used for https://github.com/explore 
- [Fluid](https://fluid.style/docs/components/buttons) used for https://github.com/features
- [Gloss](https://gloss.githubapp.com/?path=/story/components-button-button--primary) not used yet? 

This PR tries to replace the "Primer CSS Marketing" buttons with "Fluid" buttons.

- Is the long term plan to replace all of "Primer CSS Marketing" with "Fluid"? Or just the buttons?
- Will "Gloss" eventually replace "Fluid"?
- Are the `rem` based units used for the buttons dependent on [html-fluid](https://github.com/github/github/blob/215db0ae34321a0956ce0396ec1ae85bd472e5d3/app/assets/stylesheets/marketing/utilities/base.scss#L3). I think the font-size scales down on smaller screens.

Here some random pages that would probably be affected of this change:

Example | Before | After
--- | --- | ---
https://github.com/explore | ![Screen Shot 2021-04-21 at 14 22 47](https://user-images.githubusercontent.com/378023/115502938-a9b3aa80-a2b0-11eb-87f9-4055c6d954d1.png) | ![Screen Shot 2021-04-21 at 14 22 59](https://user-images.githubusercontent.com/378023/115502941-aa4c4100-a2b0-11eb-801e-07bde16bfa45.png)
https://github.com/marketplace | ![Screen Shot 2021-04-21 at 14 24 12](https://user-images.githubusercontent.com/378023/115502944-aae4d780-a2b0-11eb-8c99-f4a0fa3945ff.png) | ![Screen Shot 2021-04-21 at 14 24 22](https://user-images.githubusercontent.com/378023/115502947-ab7d6e00-a2b0-11eb-80d3-c756dd37f34f.png)
https://github.com/customer-stories | ![Screen Shot 2021-04-21 at 14 25 16](https://user-images.githubusercontent.com/378023/115502948-ac160480-a2b0-11eb-9230-484ddc22c3cb.png) | ![Screen Shot 2021-04-21 at 14 25 26](https://user-images.githubusercontent.com/378023/115502951-ac160480-a2b0-11eb-82ad-7d530dfb59c7.png)
https://github.com/security/incident-response | ![Screen Shot 2021-04-21 at 14 00 48](https://user-images.githubusercontent.com/378023/115502926-a6b8ba00-a2b0-11eb-9c8b-74bf25795408.png) | ![Screen Shot 2021-04-21 at 14 01 01](https://user-images.githubusercontent.com/378023/115502935-a91b1400-a2b0-11eb-8c87-754b8555a1b4.png)

Is that ok? Only tested it with some user-style hack, so might not be super accurate.

tobiasahlin commented:
> Is the long term plan to replace all of "Primer CSS Marketing" with "Fluid"? Or just the buttons?

Not just the buttons, we're merging any overlap between the two. Tracking issue is here https://github.com/github/site-design/issues/1190. It's part of our initiative to [detangle our CSS tooling](https://github.com/github/site-design/issues/1194) for marketing sites. The goal is to remove the `-fluid` namespace completely.

> Will "Gloss" eventually replace "Fluid"?

Yes 🙏 by further updating Primer Marketing

> Here some random pages that would probably be affected by this change:

I think these mostly look OK! I've created a PR here https://github.com/github/github/pull/177537 to complete this transition with some more polish, to make sure that this PR in this repo doesn't produce any visual changes when pulled into .com..

tobiasahlin commented:
@simurai I've updated these to support color modes, and moved all variables to `primitives`. PR for that available here: https://github.com/primer/primitives/pull/76

simurai commented on src/marketing/buttons/button.scss:
```diff
@@ -1,117 +1,106 @@
+
+  @include btn-solid-mktg(
+    var(--color-text-white),
+    lighten($primary-blue-mktg, 5%),
+    $primary-blue-mktg,
+    lighten($secondary-blue-mktg, 5%),
+    $secondary-blue-mktg
```
I think this could also be replaced with variables from primitives. Then the SASS variables [here](https://github.com/primer/css/pull/1352/files#diff-cb94cf5afa0ea8c52869cc8fcbf1c89736a5c69629624347f62b924da09ac590) shouldn't be needed anymore.

tobiasahlin replied:
Great point! I fixed this instance now ✨

simurai reviewed (approved):
Ok, I think this looks good. 👍 

Eventually we'll have to figure what to do with the Alliance font, see https://github.com/primer/css/pull/1352#pullrequestreview-639085844, but we can follow up with that on another PR.
